src/wp-includes/classicpress/class-wp-compat.php

```diff
@@ -0,0 +1,310 @@
+		if ( $this->blocks_compatibility_level !== 2 ) {
+			return;
+		}
+		$trace = debug_backtrace();
+		if ( strpos( $trace[1]['file'], get_stylesheet_directory() ) === 0 ) {
+			// Current theme is calling the function
+			update_option( 'theme_using_blocks', '1' );
+		} elseif ( strpos( $trace[1]['file'], get_template_directory() ) === 0 ) {
+			// Parent theme is calling the function
+			update_option( 'theme_using_blocks', '2' );
+		} else {
+			// A plugin is calling the function
+			$traces = array_column( $trace, 'file' );
+			$traces = array_map(
+				function( $path ) {
+					return $this->plugin_folder( $path );
+				},
+				$traces
+			);
+			$active = wp_get_active_and_valid_plugins();
+			$active = array_map(
+				function( $path ) {
+					return $this->plugin_folder( $path );
+				},
+				$active
+			);
+			$plugins = array_intersect( $traces, $active );
+			$plugin = array_pop( $plugins );
+			if ( $plugin === null ) {
+				// Nothing found? Bail.
+				return;
+			}
+			$plugins_using_blocks = get_option( 'plugins_using_blocks', array() );
+			if ( ! array_key_exists( plugin_basename( $plugin ), $plugins_using_blocks ) ) {
+				$plugins_using_blocks[ plugin_basename( $plugin ) ] = true;
+				update_option( 'plugins_using_blocks', $plugins_using_blocks );
+			}
+		}
+	}
+
+	/**
+	 * Polyfills that can not be defined elsewhere go here.
+	 * that may not work properly.
+	 *
+	 * @return void
+	 */
+	private function define_polyfills() {
+		// Polyfills that must not have to be defined elsewhere go there.
+		if ( ! function_exists( 'register_block_type' ) ) :
+			/**
+			 * Polyfill for block functions.
+			 *
+			 * @since CP-2.0.0
+			 *
+			 * @return bool False.
+			 */
+			function register_block_type( ...$args ) {
+				global $wp_compat;
+				$wp_compat->using_block_function();
+				return false;
+			}
+
+		endif;
+
+		if ( ! function_exists( 'register_block_type_from_metadata' ) ) :
+			/**
+			 * Polyfill for block functions.
+			 *
+			 * @since CP-2.0.0
+			 *
+			 * @return bool False.
+			 */
+			function register_block_type_from_metadata( ...$args ) {
+				global $wp_compat;
+				$wp_compat->using_block_function();
+				return false;
+			}
+
+		endif;
+
+		if ( ! function_exists( 'has_block' ) ) :
+			/**
+			 * Polyfill for block functions.
+			 *
+			 * @since CP-2.0.0
+			 *
+			 * @return bool False.
+			 */
+			function has_block( ...$args ) {
+				global $wp_compat;
+				$wp_compat->using_block_function();
+				return false;
+			}
+
+		endif;
+
+		if ( ! function_exists( 'register_block_pattern' ) ) :
+			/**
+			 * Polyfill for block functions.
+			 *
+			 * @since CP-2.0.0
+			 *
+			 * @return bool False.
+			 */
+			function register_block_pattern( ...$args ) {
+				global $wp_compat;
+				$wp_compat->using_block_function();
+				return false;
+			}
+
+		endif;
+
+		// Load classes files as classes can not be defined inside a class.
+		require_once 'class-wp-block-type-pf.php';
```

Comment:
For consistency through the code we should pass the full patch using ABSPATH etc as needed.

`require_once ABSPATH . WPINC . '/classicpress/class-wp-block-type-pf.php';`